### PR TITLE
Perf optimizations with function ID resolution

### PR DESCRIPTION
**Pull request by fabiocav**
Perf optimizations when resolving function IDs.

For the method optimization alone, these are the high level results:

Case	ResultsGraph	Mean	Min	Max	Range	    AllocatedBytes	Operations	Phase
TestNew		14.90 ns	14.58 ns	15.31 ns	5%	1,544	503,316,480	Complete
TestOld		29.29 ns	28.85 ns	30.14 ns	4%	1,652	234,881,024	Complete

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs [hunk 1]
```diff
@@ -741,7 +742,11 @@ internal FunctionEnvironmentReloadRequest GetFunctionEnvironmentReloadRequest(ID
         internal void SendFunctionLoadRequest(FunctionMetadata metadata, ManagedDependencyOptions managedDependencyOptions)
         {
             _functionLoadRequestResponseEvent = _metricsLogger.LatencyEvent(MetricEventNames.FunctionLoadRequestResponse);
-            _workerChannelLogger.LogDebug("Sending FunctionLoadRequest for function: '{functionName}' with functionId: '{functionId}'", metadata.Name, metadata.GetFunctionId());
+
+            if (_workerChannelLogger.IsEnabled(LogLevel.Debug))
```
Comment (kshyju) [inline src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:746]: Is it a valid case where we will disable debug level log?
We have other `LogDebug` calls in the cold start path. Should all of them be fixed?

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs [hunk 2]
```diff
@@ -406,10 +406,11 @@ public async Task InvokeAsync(ScriptInvocationContext invocationContext)
             // This could throw if no initialized workers are found. Shut down instance and retry.
             IEnumerable<IRpcWorkerChannel> workerChannels = await GetInitializedWorkerChannelsAsync(invocationContext.FunctionMetadata.Language ?? _workerRuntime);
             var rpcWorkerChannel = _functionDispatcherLoadBalancer.GetLanguageWorkerChannel(workerChannels);
-            if (rpcWorkerChannel.FunctionInputBuffers.TryGetValue(invocationContext.FunctionMetadata.GetFunctionId(), out BufferBlock<ScriptInvocationContext> bufferBlock))
+            string functionId = invocationContext.FunctionMetadata.GetFunctionId();
```
Comment (kshyju) [inline src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:409]: Can you update line 203 to 
```csharp
&& webhostLanguageWorkerChannel.Any(a => a.Value.Task.Status == TaskStatus.RanToCompletion))
```

Replacing Where + Any to just Any.

Comment (kshyju) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:409]: Also 547 to 
```csharp
if (!(await GetInitializedWorkerChannelsAsync()).Any())
```

**Review by kshyju — APPROVED** [review]

File: src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs (hunk 1 above)
Comment (fabiocav) [reply src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:746]: Yes, it is valid. Even enabled by default in hosted environments, running the host in different contexts (e.g., self-hosted, custom containers) should support this.

We should do the same for other calls, but I want to use a better pattern that is both, more performant and cleaner. Will have that as a separate change, but wanted to get this one in.

File: src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs (hunk 2 above)
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:409]: Made those changes in addition to a couple of other improvements I was tracking.

**Review by liliankasem — APPROVED** [review]

**Pull request merged**